### ISSUE TITLE
reconcile requesting 'any' and SCS selection of VLAN tags

### ISSUE THREAD

**Issue by ahelsing** [GENI-NSF/geni-tools#650]
Stitcher now changes the requested tag to 'any' to improve success.

But the SCS carefully edits the availRange so that path2 excludes the tag they suggested for path1.

The problem is when an AM is picking 'any' for multiple links, the AM might pick the tag set aside for path2 on path1, causing the next AM in the chain to complain that there is no intersection between what the SCS told it to use and what the previous AM used.

Note that if the AM honored the avail range picked by the SCS in the first place, there would be no problem, but it still might not be optimal.

 _Imported from trac ticket [#650](http://trac.gpolab.bbn.com/gcf/ticket/650),  created by ahelsing on 06-18-2014 at 14:43, last modified: 05-15-2015 at 08:55_

**Comment by ahelsing** [comment]
Expanding on the comment in the description: If AMs honor the range_request, then if there are N paths using the same hop, there are N-1 tags artificially excluded from each range. Where an AM has a limited number of VLAN tags, this might make a difference.


_Trac comment by ahelsing on 10-07-2014 at 16:30_

**Comment by ahelsing** [comment]
To re-iterate the problem: SCS excludes other paths tags from this paths range_request to avoid duplicate reservation. But when I change suggested to 'any' allowing the AM to pick a different one, I'm still excluding a tag that might turn out to be available. And so I might be artificially limiting my choices enough that the request fails.

I assume an AM where we request 'any' does not depend on another AM.

We would solve the problem if the AMs were smart (maybe they are) and we allowed the AMs to pick tags on all paths/hops simultaneously, trusting the AM to avoid duplicate provisioning of VLAN tags.

So if the AMs can avoid duplicate allocations, then

If we used 'any' at all in this topology:
 - walk dependency tree in reverse (as it comes from SCS). That is, start with things that import from others, or that neither import nor are imported from
 - Get the latest Ad with --available (using that info if available)
 - Set the range_request
 - Update that range_request by intersecting with the hop that imports from it, if any

Basically, if we use 'any', we'd throw away the SCS avail range values, and recreate them based on current Ads intersected up the chains of dependency.

But again, do AMs do the right thing in the multi circuit case?

I _believe_ it is not enough to just get the latest Ad for the 'any' hops (AMs?) because the SCS may have excluded the picked tags on the other path at other AMs in the path, so we'd just end up re-excluding those.

Hmm. What about the root of a path that doesn't use 'any'? I think I don't want to update that from what the SCS said.


_Trac comment by ahelsing on 12-11-2014 at 14:42_
